File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/main/MainActivity.java
package ru.nikshlykov.englishwordsapp.ui.main;

import android.content.Intent;
import android.os.Bundle;

import com.google.android.material.bottomnavigation.BottomNavigationView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.NavigationUI;

import ru.nikshlykov.englishwordsapp.App;
import ru.nikshlykov.englishwordsapp.R;
import ru.nikshlykov.englishwordsapp.ui.modes.ModesActivity;
import ru.nikshlykov.englishwordsapp.ui.settings.SettingsActivity;

import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity
        implements ProfileFragment.ProfileFragmentReportListener {

    // Тег для логирования.
    private static final String LOG_TAG = "MainActivity";

    // Коды запросов для общения между Activity.
    private final static int REQUEST_CODE_EDIT_MODES = 1;
    private final static int REQUEST_CODE_EDIT_SETTINGS = 2;

    // View элементы.
    private BottomNavigationView bottomNavigationView; // Нижнее меню.

    // Время последнего нажатия на кнопку "Назад" в данном Activity.
    private static long lastBackPressedTime;

    /*private BottomNavigationView.OnNavigationItemSelectedListener mOnNavigationItemSelectedListener
            = new BottomNavigationView.OnNavigationItemSelectedListener() {
        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            Fragment fragment;
            FragmentManager fragmentManager = getSupportFragmentManager();

            switch (item.getItemId()) {
                case R.id.activity_main_menu___study:
                    // Пытаемся найти фрагмент и проверяем, создан ли он (на экране).
                    fragment = fragmentManager.findFragmentByTag(TAG_STUDY_OR_INFO_FRAGMENT);
                    // Если фрагмент не создан, тогда заменяем тот фрагмент, который на экране, только что созданным.
                    if (fragment == null) {
                        if (studyViewModel.selectedModesExist()) {
                            getNextAvailableToRepeatWord();
                        } else {
                            displayInfoFragment(InfoFragment.FLAG_MODES_ARE_NOT_CHOSEN);
                        }
                    }
                    return true;

                case R.id.activity_main_menu___groups:
                    // Пытаемся найти фрагмент и проверяем, создан ли он (на экране).
                    fragment = fragmentManager.findFragmentByTag(TAG_GROUPS_FRAGMENT);
                    // Если фрагмент не создан, тогда меняем тот фрагмент, который на экране, только что созданным.
                    if (fragment == null) {
                        // Сохраняем последний фрагмент режима.
                        fragmentManager
                                .beginTransaction()
                                .replace(contentLayoutId, new GroupsFragment(), TAG_GROUPS_FRAGMENT)
                                .commit();
                    }
                    return true;

                case R.id.activity_main_menu___profile:
                    // Пытаемся найти фрагмент и проверяем, создан ли он (на экране).
                    fragment = fragmentManager.findFragmentByTag(TAG_PROFILE_FRAGMENT);
                    // Если фрагмент не создан, тогда меняем тот фрагмент, который на экране, только что созданным.
                    if (fragment == null) {
                        // Сохраняем последний фрагмент режима.
                        fragmentManager
                                .beginTransaction()
                                .replace(contentLayoutId, new ProfileFragment(), TAG_PROFILE_FRAGMENT)
                                .commit();
                    }
                    return true;
            }
            return false;
        }
    };*/

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViews();

        // Присваиваем обработчик нажатия на нижнее меню.
        //bottomNavigationView.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener);

        NavController navController = Navigation.findNavController(this, R.id.activity_main___nav_host_fragment);
        NavigationUI.setupWithNavController(bottomNavigationView, navController);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        ((App) getApplicationContext()).getTextToSpeech().shutdown();
    }

    /**
     * Находит View элементы в разметке.
     */
    private void findViews() {
        bottomNavigationView = findViewById(R.id.navigation);
    }

    // TODO Решить, что делать с кодом общения с другими Activity.

    // Общение с другими Activity и Fragments, находящимися внутри

    /**
     * Открывает Activity режимов, чтобы получить от него результат - выбранные режимы.
     */
    @Override
    public void reportOpenModesActivity() {
        Log.i(LOG_TAG, "reportOpenModesActivity()");
        Intent intent = new Intent(this, ModesActivity.class);
        startActivityForResult(intent, REQUEST_CODE_EDIT_MODES);
    }

    /**
     * Открывает Activity настроек, чтобы получить от него результат - максимальное количество
     * новых слов в день.
     */
    @Override
    public void reportOpenSettingsActivity() {
        Log.i(LOG_TAG, "reportOpenSettingsActivity()");
        Intent intent = new Intent(this, SettingsActivity.class);
        startActivityForResult(intent, REQUEST_CODE_EDIT_SETTINGS);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == REQUEST_CODE_EDIT_MODES) {
                // Достаём пришедший массив id выбранных режимов.
                Bundle extras = data.getExtras();
                long[] selectedModesIds = extras.getLongArray(ModesActivity.EXTRA_SELECTED_MODES);

                // Преобразовываем его в ArrayList.
                ArrayList<Long> selectedModesIdsList = new ArrayList<>();
                Log.i(LOG_TAG, "selectedModesIds: ");
                for (int i = 0; i < selectedModesIds.length; i++) {
                    selectedModesIdsList.add(selectedModesIds[i]);
                    Log.i(LOG_TAG, "mode id: " + selectedModesIds[i]);
                }

                // Закидываем в studyViewModel для хранения.
                //studyViewModel.setSelectedModesIds(selectedModesIdsList);
            } else if (requestCode == REQUEST_CODE_EDIT_SETTINGS) {
                // Закидываем в studyViewModel новое количество новых слов в день.
                //studyViewModel.setNewWordsCount(data.getExtras().getInt(SettingsActivity.EXTRA_MAX_WORD_COUNT));
            }
        }
    }


    // Обработка выхода из приложения.

    /**
     * Обрабатывает нажатие кнопки назад и информирует о том, что
     * необходимо нажать её два раза для выхода из приложения.
     */
    @Override
    public void onBackPressed() {
        if (lastBackPressedTime + 2000 > System.currentTimeMillis())
            super.onBackPressed();
        else
            Toast.makeText(this, "Нажмите ещё раз для выхода!",
                    Toast.LENGTH_SHORT).show();
        lastBackPressedTime = System.currentTimeMillis();
    }
}

File: app/src/main/java/ru/nikshlykov/englishwordsapp/ui/MainActivity.kt
package ru.nikshlykov.englishwordsapp.ui

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.NavigationUI
import com.google.android.material.bottomnavigation.BottomNavigationView
import ru.nikshlykov.englishwordsapp.App
import ru.nikshlykov.englishwordsapp.NavigationMainDirections
import ru.nikshlykov.englishwordsapp.R
import ru.nikshlykov.englishwordsapp.ui.fragments.OnBoardingRouter
import ru.nikshlykov.feature_groups_and_words.ui.flowfragments.GroupsAndWordsFlowFragment
import ru.nikshlykov.feature_profile.navigation.ProfileFeatureRouter
import ru.nikshlykov.feature_profile.ui.flowfragments.ProfileFlowFragment
import ru.nikshlykov.feature_profile.ui.flowfragments.ProfileFlowFragmentDirections
import ru.nikshlykov.feature_study.navigation.StudyFeatureRouter
import ru.nikshlykov.feature_study.ui.flowfragments.StudyFlowFragment
import ru.nikshlykov.feature_study.ui.flowfragments.StudyFlowFragmentDirections

class MainActivity : AppCompatActivity(), ProfileFeatureRouter, StudyFeatureRouter,
  OnBoardingRouter {

  // TODO посмотреть, что вообще с flowfragments. у них можно поставить internal или нет.

  // TODO проверить все вьюхи с клавой на adjustresize.

  // TODO сделать скрытие bottomNavigation, когда у нас не корневые фрагменты.
  //  Иначе не сохраняется тот фрагмент, на котором мы были. Конечно, лучше, чтобы
  //  сохранялся, но пока можно не реализовывать для простоты.

  // TODO feature. Сделать переход на начальные экраны вкладок bottomview по тапу на вкладки
  //  (изучение, группы, профиль).

  // TODO fix. Проблема с BottomNavView. Если в первом flow уйти далеко, потом перейти на второй,
  //  а потом вернуться на первый, то подсвечивается иконка второго, а не первого.

  // TODO feature. Сделать удержание не только фрагментов во flow, но и позиций списков, чтобы ui
  //  полностью оставался таким же при возвращении.
  private var navHostFragment: NavHostFragment? = null

  private var navController: NavController = NavController(this)

  private lateinit var bottomNavigationView: BottomNavigationView

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    installSplashScreen()
    (applicationContext as App).mainActivity = this
    setContentView(R.layout.activity_main)
    bottomNavigationView = findViewById(R.id.navigation)
    navHostFragment =
      supportFragmentManager.findFragmentById(R.id.activity_main___nav_host_fragment) as NavHostFragment
    navHostFragment?.navController?.let {
      this.navController = it
    }
    NavigationUI.setupWithNavController(bottomNavigationView, navController)

    if (!checkOnBoardingIsFinished()) {
      bottomNavigationView.visibility = View.GONE
      navController.navigate(NavigationMainDirections.actionGlobalOnBoardingViewPagerDest())
    }
  }

  override fun onDestroy() {
    super.onDestroy()
    (applicationContext as App).textToSpeech.shutdown()
    // TODO refactor. Нужно ли занулять подобным образом dagger component?
  }

  // Для обработки выхода из приложения.
  override fun onBackPressed() {
    // TODO разобраться потом с тем, как выходить из любой вкладки (что Profile, что GroupsAndWords)
    val primaryFragment = navHostFragment?.childFragmentManager?.primaryNavigationFragment
    if (primaryFragment is GroupsAndWordsFlowFragment && primaryFragment.backPressedIsAvailable()) {
      Log.d(LOG_TAG, "GroupsAndWordsFlowFragment onBackPressed()")
      primaryFragment.onBackPressed()
    } else if (primaryFragment is ProfileFlowFragment && primaryFragment.backPressedIsAvailable()) {
      Log.d(LOG_TAG, "ProfileFlowFragment onBackPressed()")
      primaryFragment.onBackPressed()
    } else {
      if (primaryFragment is StudyFlowFragment) {
        if (lastBackPressedTime + 2000 > System.currentTimeMillis()) {
          Log.d(LOG_TAG, "super.onBackPressed() if StudyFlowFragment")
          super.onBackPressed()
        } else {
          Toast.makeText(
            this, "Нажмите ещё раз для выхода!",
            Toast.LENGTH_SHORT
          ).show()
          lastBackPressedTime = System.currentTimeMillis()
        }
      } else {
        super.onBackPressed()
      }
    }
  }

  override fun openStatistics() {
    val navDirections = ProfileFlowFragmentDirections.actionProfileFlowDestToStatisticsDest()
    navController.navigate(navDirections)
  }

  override fun openModes() {
    val navDirections = ProfileFlowFragmentDirections.actionProfileFlowDestToModesDest()
    navController.navigate(navDirections)
  }

  override fun openSettings() {
    val navDirections = ProfileFlowFragmentDirections.actionProfileFlowDestToSettingsDest()
    navController.navigate(navDirections)
  }

  override fun openModesFromStudy() {
    val navDirections = StudyFlowFragmentDirections.actionStudyFlowDestToModesDest()
    navController.navigate(navDirections)
  }

  override fun openGroupsFromStudy() {
    val navDirections = StudyFlowFragmentDirections.actionStudyFlowDestToGroupsAndWordsFlowDest()
    navController.navigate(navDirections)
  }

  private fun checkOnBoardingIsFinished(): Boolean {
    val sharedPref = getSharedPreferences(ON_BOARDING_PREFERENCES_NAME, Context.MODE_PRIVATE)
    return sharedPref.getBoolean("Finished", false)
  }

  override fun endOfOnBoarding() {
    navController.popBackStack()
    bottomNavigationView.visibility = View.VISIBLE
    val sharedPref = getSharedPreferences(ON_BOARDING_PREFERENCES_NAME, Context.MODE_PRIVATE)
    val editor = sharedPref.edit()
    editor.putBoolean("Finished", true)
    editor.apply()
  }

  companion object {
    private const val LOG_TAG = "MainActivity"

    private var lastBackPressedTime: Long = 0
    private const val ON_BOARDING_PREFERENCES_NAME = "OnBoarding"
  }
}